sylt-compiler/src/dependency.rs

```diff
@@ -0,0 +1,298 @@
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+use std::collections::btree_map::Entry::{Occupied, Vacant};
+use crate::{Compiler, Name};
+use sylt_parser::{
+    AST, Assignable, AssignableKind, Expression, ExpressionKind, Identifier,
+    Statement, StatementKind,
+};
+use sylt_parser::statement::NameIdentifier;
+
+struct Context<'a> {
+    compiler: &'a Compiler,
+    namespace: usize,
+    variables: Vec<String>,
+}
+
+impl Context<'_> {
+    fn shadow(&mut self, variable: &String) {
+        if !self.shadowed(variable) {
+            self.variables.push(variable.clone());
+        }
+    }
+
+    fn shadowed(&self, variable: &String) -> bool {
+        return self.variables.iter().rfind(|&v| v == variable).is_some();
+    }
+}
+
+
+
+fn assignable_dependencies(ctx: &mut Context, assignable: &Assignable) -> BTreeSet<Name> {
+    use AssignableKind::*;
+    match &assignable.kind {
+        Read(ident) => {
+            match ctx.compiler.namespaces[ctx.namespace].get(&ident.name) {
+                Some(&name) if !ctx.shadowed(&ident.name) => {
+                    [name].iter().cloned().collect()
+                },
```

Comment:
I don't really get this but it looks weird.

Reply:
It is how you create a set without a variable.

Reply:
I see. In that case, I think I'd want it to be a bit more explicit. Maybe `[name.clone].into_iter().collect::<BTreeSet<_>>()` (untested, sorry).